File: problem_032.c
// -*- compile-command: "gcc -std=c99 -o problem_032 problem_032.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

bool isPandigitalProduct(int a, int b, int c)
{
    // Create check array for digits. Mark 0 as 1 since 0 not included in pandigital numbers
    int check[10] = {0};
    check[0] = 1;

    while (a != 0)
    {
        check[a%10]++;
        a /= 10;
    }

    while (b != 0)
    {
        check[b%10]++;
        b /= 10;
    }

    while (c != 0)
    {
        check[c%10]++;
        c /= 10;
    }

    for (int i=0; i<10; i++)
        if (check[i] != 1)
            return false;

    return true;
}

int main()
{
    printf("Project Euler - Problem 32:\n"
           "Find the sum of all numbers that can be written as pandigital products.\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int sum = 0;
    int found[100] = {};
    int count = 0;
    bool debug = false;

    for (int a=1 ; a<2500; a++)
        for (int b=a ; b<2500; b++)
        {
            int c = a*b;

            // If pandigital, record, increment
            if (isPandigitalProduct(a, b, c) == true)
            {
                // Check history for c
                bool recorded = false;
                for (int i=0 ; i<count ; i++)
                    if (found[i] == c)
                        recorded = true; //Already recorded, ignore

                //Record and add to sum
                if (recorded == false)
                {
                    if (debug == true) printf("%d x %d = %d\n", a, b, c);
                    found[count++]=c;
                    sum += c;
                }
            }
        }

    printf("%d\n", sum);
    printElapsedTime(start);

    return 0;
}

File: problem_032.rs
// -*- compile-command: "rustc -o problem_032_rs problem_032.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn is_pandigital(a: usize, b: usize, c: usize) -> bool {
    let mut a = a;
    let mut b = b;
    let mut c = c;
    let mut key = [1, 0, 0, 0, 0, 0, 0, 0, 0, 0];

    while a > 0 {
        key[a%10] += 1;
        a /= 10;
    }

    while b > 0 {
        key[b%10] += 1;
        b /= 10;
    }

    while c > 0 {
        key[c%10] += 1;
        c /= 10;
    }

    for i in 0..key.len() {
        if key[i] != 1 {
            return false;
        }
    }

    return true;
}

fn main() {
    println!("Project Euler - Problem 32:");
    println!("Find the sum of all numbers that can be written as pandigital products.\n");

    let mut sum = 0;
    let mut products = Vec::new();

    for a in 0..99 {
        for b in a..5000 {
            let c = a * b;
            if is_pandigital(a, b, c) == true {
                let mut new = true;
                for i in 0..products.len() {
                    if products[i] == c {
                        new = false;
                    }
                }
                if new == true {
                    sum += c;
                    products.push(c);
                    //println!("{} x {} = {}", a, b, c); // DEBUG
                }
            }
        }
    }

    println!("{}", sum);
}
